File: concepts/07_modules_packages_and_loading/01_modules_imports_linkage_and_live_bindings/julia/test_02_binding_aliases_and_mutable_exports.jl
# polyglot-family: modules_packages_and_loading
# polyglot-concept: modules_imports_linkage_and_live_bindings
# polyglot-related: languages/julia/tooling_and_runtime/11_modules_macros_and_metaprogramming/
# polyglot-related+: test_081_module_namespaces_exports_and_qualification.jl
#
# 共同问题：导入别名能否重新绑定；导出可变对象的内部修改是否对所有引用可见。
# 对照观察：imported binding 只读，调用方可建立自己的 const alias；对象 mutation 与 binding 重绑定分离。

using Test

module AliasSource
export state
const state = Dict(:count => 0)
end

const local_state_alias = AliasSource.state

@testset "alias 共享对象但拥有独立名称" begin
    local_state_alias[:count] += 1
    @test AliasSource.state[:count] == 1
    @test local_state_alias === AliasSource.state
    @test isconst(AliasSource, :state)
end

File: concepts/07_modules_packages_and_loading/01_modules_imports_linkage_and_live_bindings/rust/test_02_binding_aliases_and_mutable_exports.rs
// polyglot-family: modules_packages_and_loading
// polyglot-concept: modules_imports_linkage_and_live_bindings
// polyglot-related: languages/rust/tests/tooling_and_runtime/
// polyglot-related+: 10_modules_visibility_linkage_and_initialization/test_073_modules_visibility_and_reexports.rs
//
// 共同问题：import alias 是否创建新模块实例；可变导出是否共享；局部重绑定影响谁。
// 对照观察：`use ... as` 只重命名本地 item binding；两个 alias 指向同一 static，不执行第二次初始化。

use std::sync::atomic::{AtomicUsize, Ordering};

static SHARED: AtomicUsize = AtomicUsize::new(0);
use SHARED as FIRST;
use SHARED as SECOND;

#[test]
fn comparison() {
    FIRST.store(42, Ordering::SeqCst);
    assert_eq!(SECOND.load(Ordering::SeqCst), 42);
    let snapshot = SECOND.load(Ordering::SeqCst);
    FIRST.store(7, Ordering::SeqCst);
    assert_eq!(snapshot, 42);
    assert_eq!(SECOND.load(Ordering::SeqCst), 7);
}
